tensorflow/compiler/mlir/xla/transforms/xla_hlo_fusion.cc

```diff
@@ -0,0 +1,553 @@
+  bool TryToContractEdge(Cluster* from, Cluster* to) {
+    int node_to = to->cycles_graph_node_id();
+    int node_from = from->cycles_graph_node_id();
+
+    // Both node_to and node_from should be fusible
+    if (!IsFusible(op_list_[node_to]) || !IsFusible(op_list_[node_from])) {
+      return false;
+    }
+
+    auto op_from_fusibility = dyn_cast<InferFusibilityOpInterface>(op_list_[node_from]);
+    if (op_from_fusibility && !op_from_fusibility.isFusibleWithConsumer()) {
+      // This op cannot be fused with its consumers.
+      return false;
+    }
+
+    auto op_to_fusibility = dyn_cast<InferFusibilityOpInterface>(op_list_[node_to]);
+    if (op_to_fusibility && !op_to_fusibility.isFusibleWithOperand()) {
+      // This op cannot be fused with its operands.
+      return false;
+    }
+
+    // Output shapes of a fusion pattern should be compatible as described in
+    // the document of this class.
+    SmallVector<Value, 4> results = GetResultsOfFusedPattern(from, to);
+    auto get_workload_shape = [] (Value v) {
+      Operation* op = v.getDefiningOp();
+      // Block argument
+      if (!op) return v;
+      auto op_fusibility = dyn_cast<InferFusibilityOpInterface>(op);
+      // Const value
+      if (!op_fusibility) return v;
+      llvm::Optional<Value> workload = op_fusibility.inferEffectiveWorkloadShape();
+      return workload.hasValue() ? *workload : v;
+    };
+
+    Value ref = get_workload_shape(results[0]);
+    if (!llvm::all_of(results, [&] (Value result) {
+      Value val = get_workload_shape(result);
+      return shape_analysis_.HasSameShape(ref, val);
+    })) {
+      return false;
+    }
+
+    return MergeClusters(from, to);
+  }
+
+  // Greedily fuse connected node.
+  bool RunEdgeContractionLoop() {
+    using std::placeholders::_1;
+    using std::placeholders::_2;
+    return ForEachEdgeInPostOrder(
+        std::bind(&FusionPlanner::TryToContractEdge, this, _1, _2));
+  }
+
+  const SmallVectorImpl<Operation *>& op_list_;
+
+  // Shape equality checker
+  ShapeConstraintAnalysis shape_analysis_;
+
+  // op -> node_id
+  std::unordered_map<Operation*, int> op_to_node_id_;
+
+  // make sure not introduce cycle after fusion
+  GraphCycles cycle_detector_;
+  std::vector<std::unique_ptr<Cluster>> cluster_storage_;
+
+  // a UnionFind set
+  EquivalenceClasses<int32_t> leader_for_node_;
+};
+
+struct XlaHloFusion : public mlir::PassWrapper<XlaHloFusion, FunctionPass> {
+  void runOnFunction() override {
+    FuncOp func = getFunction();
+    if (!IsTargetFunc(func)) {
+      return;
+    }
+
+    // process each block and do fusion within a block.
+    for (Block& block : func.getBlocks()) {
+      SmallVector<Operation*, 4> op_list;
+      for (Operation& op : block) {
+        op_list.push_back(&op);
+      }
+
+      FusionPlanner planner(op_list);
+      llvm::Optional<FusionPlan> plan = planner.Run();
+      if (!plan) {
+        emitError(func.getLoc(), "Fusion planner failed.");
```

Comment:
Errors emitted should be sentence fragments, starting with lower case and not terminated by punctuation